fwdctl/src/dump.rs

```diff
@@ -24,50 +28,230 @@ pub struct Options {
     let db = Db::new(opts.db.clone(), cfg.build()).await?;
     let latest_hash = db.root_hash().await?;
     let Some(latest_hash) = latest_hash else {
         println!("Database is empty");
         return Ok(());
     };
     let latest_rev = db.revision(latest_hash).await?;
-    let start_key = opts.start_key.clone().unwrap_or(Box::new([]));
+
+    let start_key = opts.start_key.clone().unwrap_or_else(|| Box::new([]));
+    let stop_key = opts.stop_key.clone().unwrap_or_else(|| Box::new([]));
+    let mut key_count = 0;
+
     let mut stream = MerkleKeyValueStream::from_key(&latest_rev, start_key);
-    loop {
-        match stream.next().await {
-            None => break,
-            Some(Ok((key, value))) => {
-                if opts.hex {
-                    println!("'{}': '{}'", hex::encode(&key), hex::encode(&value));
-                } else {
-                    println!("'{}': '{}'", u8_to_string(&key), u8_to_string(&value));
+    let mut output_handler = create_output_handler(opts).expect("Error creating output handler");
+
+    while let Some(item) = stream.next().await {
+        match item {
+            Ok((key, value)) => {
+                output_handler.handle_record(&key, &value);
+
+                key_count += 1;
+
+                if key == stop_key || key_count_exceeded(opts.max_key_count, key_count) {
+                    handle_next_key(stream.next().await).await;
+                    break;
                 }
             }
-            Some(Err(e)) => return Err(e),
+            Err(e) => return Err(e),
         }
     }
+    output_handler.flush();
+
     Ok(())
 }
 
-fn u8_to_string(data: &[u8]) -> Cow<'_, str> {
-    String::from_utf8_lossy(data)
+const fn key_count_exceeded(max: Option<u32>, key_count: u32) -> bool {
+    matches!(max, Some(max) if key_count >= max)
+}
+
+fn u8_to_string(data: &[u8]) -> String {
+    String::from_utf8_lossy(data).to_string()
 }
 
 fn key_parser(s: &str) -> Result<Box<[u8]>, std::io::Error> {
     Ok(Box::from(s.as_bytes()))
 }
+
+// Helper function to convert key and value to a string
+fn key_value_to_string(key: &[u8], value: &[u8], hex: bool) -> (String, String) {
+    let key_str = if hex {
+        hex::encode(key)
+    } else {
+        u8_to_string(key)
+    };
+    let value_str = if hex {
+        hex::encode(value)
+    } else {
+        u8_to_string(value)
+    };
+    (key_str, value_str)
+}
+
+async fn handle_next_key(next_key: KeyFromStream) {
+    match next_key {
+        Some(Ok((key, _))) => {
+            println!(
+                "Next key is {0}, resume with \"--start-key={0}\".",
+                u8_to_string(&key)
+            );
+        }
+        Some(Err(e)) => {
+            eprintln!("Error occurred while fetching the next key: {}", e);
+        }
+        None => {
+            println!("There is no next key. Data dump completed.");
+        }
+    }
+}
+
+trait OutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]);
```

Comment:
This must handle errors better here rather than ignore them. This is likely to be important for things like running out of disk space when dumping the output file.

With the way the code is now, if you run out of disk space, no errors will be reported in some cases.

```suggestion
    fn handle_record(&mut self, key: &[u8], value: &[u8]) -> Result<(), std::io::Error>
```

Flush needs this too.

You can choose to use a `dyn std::error::Error` or just assume we only handle IO errors more gracefully.

Reply:
Very much agreed, adding error handling for this